async/server: Fix the the dropping of the server object

Tim-Zhang opened the pull request:
Break the accepting-new-connection loop correctly when the server object was dropped.

Fixes: #284

jokemanfire commented:
Sorry, it couldn't the IT is fail.
```
#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;

    pub fn is_socket_in_use(sock_path: &str) -> bool {
        use std::process::Command;

        let output = Command::new("lsof")
            .arg(sock_path)
            .output()
            .expect("Failed to execute lsof command");

        output.status.success()
    }

    #[cfg(unix)]
    #[tokio::test]
    // Add this test for test thread leak, if the caller forget to call the shutdown function.
    async fn test_server_start() {
        use std::time::Duration;

        simple_logging::log_to_stderr(LevelFilter::Trace);
        {
            let hservice = health_ttrpc::create_health(Arc::new(HealthService {}));
            utils::remove_if_sock_exist(utils::SOCK_ADDR).unwrap();
            let mut server = Server::new()
                .bind(utils::SOCK_ADDR)
                .unwrap()
                .register_service(hservice);
            server.start().await.unwrap();
        }
        sleep(std::time::Duration::from_secs(1)).await;
        // judge utils::SOCK_ADDR if still occupied
        let addr = utils::SOCK_ADDR
            .strip_prefix("unix://")
            .expect("socket address is not expected");
        // It should be true, because the server's thread is not stopped.
        tokio::time::sleep(Duration::from_secs(1)).await;
        assert!(is_socket_in_use(addr));
    }
}
```

jokemanfire commented:
Yes , thanks, it could ,the it and ut passed. I will close that pr. You can add this it for test.
```
#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;

    pub fn is_socket_in_use(sock_path: &str) -> bool {
        use std::process::Command;

        let output = Command::new("lsof")
            .arg(sock_path)
            .output()
            .expect("Failed to execute lsof command");

        output.status.success()
    }

    #[cfg(unix)]
    #[tokio::test]
    // Add this test for test thread leak, if the caller forget to call the shutdown function.
    async fn test_server_start() {
        use std::time::Duration;

        simple_logging::log_to_stderr(LevelFilter::Trace);
        {
            let hservice = health_ttrpc::create_health(Arc::new(HealthService {}));
            utils::remove_if_sock_exist(utils::SOCK_ADDR).unwrap();
            let mut server = Server::new()
                .bind(utils::SOCK_ADDR)
                .unwrap()
                .register_service(hservice);
            server.start().await.unwrap();
            // sleep(std::time::Duration::from_secs(30)).await;
        }
        sleep(std::time::Duration::from_secs(1)).await;
        // judge utils::SOCK_ADDR if still occupied
        let addr = utils::SOCK_ADDR
            .strip_prefix("unix://")
            .expect("socket address is not expected");
        // It should not be true, because the server's thread is not stopped.
        assert!(!is_socket_in_use(addr));
    }
}
```

jokemanfire commented:
Indeed, disconnecting the stop channel during the lifecycle will end, and no end signal will be sent here. However, the sender has already been destroyed, and I forgot to consider this.

jokemanfire commented:
```
                        match fd_tx {
                            Some(tx) => {
                                if let Err(e) = tx.send(incoming).await {
                                    error!("Failed to send listener to stop channel: {:?}", e);
                                }
                            }
                            None => {
                                debug!("Stop listen channel closed");
                            }
                        }
                        break;
```
if this will looks great ?

Tim-Zhang commented:
@jokemanfire Thanks for your suggestions. I've added the ut, and added your sign-off on the commit. As for the debug log, I think it is unnecessary to add it, so I did not add it.

Thanks again for your efforts in fixing this issue.

Tim-Zhang commented:
@teawater PTAL

Tim-Zhang commented:
@wllenyj PTAL
